ZOOKEEPER-3310: Add metrics for prep processor

jhuan31 opened the pull request.

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java:
```diff
@@ -157,8 +160,8 @@ public void run() {
         ChangeRecord lastChange = null;
-        synchronized (zks.outstandingChanges) {
-            lastChange = zks.outstandingChangesForPath.get(path);
+        synchronized (zks.getOutstandingChanges()) {
```
Can you call the getter outside the synchronized block and then. Use the reference to the returned object?
It will be clearer that we are accessing the same object

jhuan31 replied:
I wasn't feeling comfortable doing this either. Even calling the getter outside is not satisfying because this object is being used in many places. How about we add the field in PrepRequestProcessor and make it final and initialize it in the constructor?

eolivelli replied:
Okay

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java:
```diff
@@ -183,10 +186,12 @@ private ChangeRecord getOutstandingChange(String path) {
-            zks.outstandingChanges.add(c);
-            zks.outstandingChangesForPath.put(c.path, c);
+    protected void addChangeRecord(ChangeRecord c) {
+        LOG.info("addChangeRecord called");
```
Is this a debug to drop?

jhuan31 replied:
yes. will do.

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java:
```diff
@@ -588,10 +593,11 @@ protected void pRequest2Txn(int type, long zxid, Request request,
-                synchronized (zks.outstandingChanges) {
-                    for (ChangeRecord c : zks.outstandingChanges) {
+                synchronized (zks.getOutstandingChanges()) {
+                    for (ChangeRecord c : zks.getOutstandingChanges()) {
```
Same as above

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java:
```diff
@@ -902,6 +909,7 @@ protected void pRequest(Request request) throws RequestProcessorException {
             }
         }
         request.zxid = zks.getZxid();
+        ServerMetrics.PREP_PROCESSOR_QUEUE_TIME.add(Time.currentElapsedTime() - request.prepQueueStartTime);
```
What happens if prepQueueStartTime stay the default value -1 ?

jhuan31 replied:
A request that has got here must have gone through processRequest(), which records prepQueueStartTime(). Can't think of a scenario that it stays -1.

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/Request.java:
```diff
@@ -78,6 +78,8 @@ public Request(long sessionId, int xid, int type, TxnHeader hdr, Record txn, lon
 
     public final long createTime = Time.currentElapsedTime();
 
+    public long prepQueueStartTime= -1;
```
Is -1 meaningful?

jhuan31 replied:
no :)

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java:
```diff
@@ -231,6 +231,14 @@ public ZooKeeperServer(FileTxnSnapLog txnLogFactory, int tickTime)
     }
 
+    public Deque<ChangeRecord> getOutstandingChanges() {
+        return outstandingChanges;
```
Can we make this field private as we are encapsulating it?

jhuan31 replied:
I've changed the unit test so it doesn't need the getter anymore. So this field stays as is. Alos reverted the code to access the field directly (which is not nice) instead of calling the getter

jhuan31 commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/PrepRequestProcessorMetricsTest.java:
```diff
@@ -0,0 +1,175 @@
+
+        prepRequestProcessor.start();
+
+        Thread.sleep(500);
```
fixed

anmolnar commented:
Merged to master branch. Thanks @jhuan31 !

eolivelli commented:
@anmolnar does master compile any more?
This patch should have been rebased because now we have changes the ServerMetrics class

anmolnar commented:
@eolivelli Ooops, no. :(
I thought it has already been rebase, because the build was green, but probably not recent.

anmolnar commented:
@jhuan31 Would you please kindly submit another PR to fix the build?

eolivelli commented:
@anmolnar we can commit a simple git revert in order to fix master if @jhuan31 does not have cycles.
Or I can send a fix patch  tomorrow

anmolnar commented:
@eolivelli yeah, probably that would be the quickest way to resolve this.

anmolnar commented:
This is done. Thanks @eolivelli for pointing out. @jhuan31 I reopen the PR, please rebase your changes. Thanks.

jhuan31 commented:
Just came back from vacation. Will rebase/clean up the PR.

Thanks,
Jie

On Sun, Apr 14, 2019 at 11:02 AM Andor Molnár <notifications@github.com>
wrote:

> Reopened #855 <https://github.com/apache/zookeeper/pull/855>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/zookeeper/pull/855#event-2275007314>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AqOFU8r5bzqbFQMNp7xppCXeLAk3Bl5qks5vg206gaJpZM4b4MIT>
> .
>

anmolnar commented:
@jhuan31 Have you finished, ready to commit?

jhuan31 commented:
yes, rebased, modified the code to use the new framework, and all tests
passed!

On Tue, Apr 16, 2019 at 2:24 AM Andor Molnár <notifications@github.com>
wrote:

> @jhuan31 <https://github.com/jhuan31> Have you finished, ready to commit?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/zookeeper/pull/855#issuecomment-483582160>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AqOFU1hF274qtTw_EHJ3lpYMnrheonLuks5vhZbAgaJpZM4b4MIT>
> .
>

anmolnar commented:
Committed, thanks @jhuan31 !

anmolnar commented:
Code doesn't compile, I reverted the patch again.

anmolnar commented:
Trying to commit again. It didn't merge it correctly, because I made a mistake with the commit script.

anmolnar commented:
@jhuan31 Looks like a new flaky test has been introduced by this patch:

https://builds.apache.org/view/S-Z/view/ZooKeeper/job/ZooKeeper-trunk/512/testReport/junit/org.apache.zookeeper.server/PrepRequestProcessorMetricsTest/testPrepRequestProcessorMetrics/

Would you please create a de-flaky jira, I'll try to look into the test today why it become flaky?
Thanks.

anmolnar commented:
Got it. The counter of processed requested (cnt_)PREP_PROCESS_TIME is updated in the `run()` method _after_ `pRequest()`.
`pRequest()` calls nextRequestProcessor.process() which is mocked to update the latch for synchronization. As a consequence the test could end before the counter is updated.

To fix this I suggest to move the update of the counter inside the pRequest() call before `nextRequestProcessor.process(). It will be more accurate anyway.

jhuan31 commented:
Thank you for looking into this. I will create a new JIRA and put the number here.

jhuan31 commented:
JIRA 3383 fixes this. Thank you!
